File: LeetCode/April Challenge/29th Day/Day_29.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=INT_MIN;
    int solve(TreeNode* root)
    {
        if(!root)
            return 0;
        
        int l= solve(root->left);
        int r= solve(root->right);
        
        ans= max({ans, root->val, l+root->val, r+root->val, l+r+root->val});
        
        return max({root->val, l+root->val, r+root->val});
    }
    int maxPathSum(TreeNode* root) {
        if(!root)
            return 0;
        solve(root);
        return ans;
    }
};

File: LeetCode/April Challenge/29th Day/Day_29.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        self.total = float('-inf')
        
        def recur(node):
            if not node:
                return 0
            left_sum = recur(node.left)
            right_sum = recur(node.right)
            
            till_max = max(node.val, node.val + max(left_sum, right_sum))
            
            self.total = max(self.total, till_max, node.val + left_sum + right_sum)
            return till_max
        
        return max(recur(root), self.total)
